sled-agent/src/instance.rs

```diff
@@ -960,6 +1041,29 @@ impl InstanceRunner {
     }
 }
 
+fn propolis_error_code(
+    log: &slog::Logger,
+    error: &PropolisClientError,
+) -> Option<PropolisErrorCode> {
+    // Is this a structured error response from the Propolis server?
+    let propolis_client::Error::ErrorResponse(ref rv) = &error else {
+        return None;
+    };
+
+    let code = rv.error_code.as_deref()?;
+    match code.parse::<PropolisErrorCode>() {
```

Comment:
This is interesting. It suggest we might want a structured error response from dropshot if we're interested in taking discrete actions depending on those values.

Reply:
Yeah, I stumbled across oxidecomputer/dropshot#39 (which you referenced above) and the related oxidecomputer/dropshot#41 way back when I originally added these error codes to Propolis (see https://github.com/oxidecomputer/propolis/pull/649#discussion_r1499913263), and I think having a first-class mechanism for structured serializable/deserializable error code types in Dropshot/Progenitor would have made all of this much nicer. 

I don't want to hold up the short term fix for sled-agent totally failing to handle these errors on getting that into Dropshot, but it's definitely the better long-term solution.